### PR TITLE
Fix for #643-Incorrect update of pin selection combo boxes

### PR DESCRIPTION
**Pull request by GioCC**
TEST VERSION, LIMITED TO `LedSegment`!
Needs more thorough testing, but seems good.
@DocMoebiuz could you have a look as soon as convenient? If the solution is approved, I can track the other devices where the modification is required.

_PS I added some small comments here and there for orientation, is it OK as a general guideline?_

### CODE REVIEW

**Comment by DocMoebiuz** [comment]
Another approach could have been to sync the setting back to the module and refresh the fields again. It looks like to me that the current approach is very invasive.

So to say it in other words. At the moment if I make a change, then switch to another device and come back, the drop-downs should be updated accordingly, right? So after leaving the current pin drop down, all we would have to do is sync the changes and re-init the panel and we should be able to re-use all existing logic.

**Comment by GioCC** [comment]
Hm, you definitely have a point there. It seems more practical indeed.
At a first glance, though, I'm not really convinced though, because:
- (main reason) calling a re-sync would basically mean to make use syncPanelWithSelectedDevice(), which is outside of the panel class responsibility and scope;
- I don't really see this approach as "invasive": a little cumbersome, yes, but what it does different from the existing procedure is basically just the added required operations which are in the first part of the "update_all()" method (the rest is just refactoring), and everything is closely contained inside the class.
I have to admit that I tend to see the "let's rebuild everything again from scratch" approach in general as a little wasteful, but then again, I'm more used to embedded development, and in C#/.NET different rules may sensibly apply.

**Comment by GioCC** [comment]
So, what about I throw a "resync" exception and have the parent catch it?

**Review by DocMoebiuz — CHANGES_REQUESTED** [review]
Thanks for all the effort. It seems to be working.

Improvements:
* It is really hard to review the changes because sometimes, some lines were only moved around, and indenting got changed - (e.g. MFAnalogPanel.cs). 
* There are also many commented lines of code. Please remove those since we have git for history. 
* Please don't put comments at the end of the line. Always better on top, before the code.
* Also for some reason it seems that the git history at least on my local main branch is not linear anymore. When you merge the main branch, do you merge rebase  or simply merge?
* Method normally are CamelCase.
* EventHandlers in C# have an "_eventname" when the Designer adds them, so that's why some methods contain an "_"

File: UI/Panels/Device/MFStepperPanel.cs [hunk 1]
```diff
@@ -1,94 +1,115 @@
...
-            }
+            pinList = Pins; // Keep pin list stored
 
-            // TODO: Complete member initialization
             this.stepper = stepper;
-            mfPin1ComboBox.SelectedValue = byte.Parse(stepper.Pin1);
-            mfPin2ComboBox.SelectedValue = byte.Parse(stepper.Pin2);
-            mfPin3ComboBox.SelectedValue = byte.Parse(stepper.Pin3);
-            mfPin4ComboBox.SelectedValue = byte.Parse(stepper.Pin4);
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin1ComboBox, Pins, stepper.Pin1);
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFStepperPanel.cs:29]: simply remove.

File: UI/Panels/Device/MFStepperPanel.cs [hunk 2]
```diff
@@ -1,94 +1,115 @@
...
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin2ComboBox, Pins, stepper.Pin2);
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin3ComboBox, Pins, stepper.Pin3);
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin4ComboBox, Pins, stepper.Pin4);
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfBtnPinComboBox, Pins, stepper.BtnPin);
+            update_lists();
 
-            mfNameTextBox.Text = stepper.Name;
+            ////if (mfPin1ComboBox.Items.Count > 4)
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFStepperPanel.cs:36]: simply remove.

File: UI/Panels/Device/MFStepperPanel.cs [hunk 3]
```diff
@@ -1,94 +1,115 @@
...
 
             initialized = true;
         }
-        
-        private void value_Changed(object sender, EventArgs e)
+
+        private void setNonPinValues()
         {
-            if (!initialized) return;
+            ////stepper.Pin1 = mfPin1ComboBox.SelectedItem.ToString();
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFStepperPanel.cs:62]: simply remove.

File: UI/Panels/Device/MFShiftRegisterPanel.cs [hunk 4]
```diff
@@ -29,52 +25,75 @@ public MFShiftRegisterPanel()
...
             textBox1.Text = shiftRegister.Name;
 
             initialized = true;
         }
 
-        private void value_Changed(object sender, EventArgs e)
+        private void setNonPinValues()
         {
-            if (!initialized) return;
+            ////shiftRegister.LatchPin = mfPin1ComboBox.SelectedItem.ToString();
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFShiftRegisterPanel.cs:58]: simply remove.

File: UI/Panels/Device/MFShiftRegisterPanel.cs [hunk 5]
```diff
@@ -29,52 +25,75 @@ public MFShiftRegisterPanel()
...
                 mfNumModulesComboBox.Items.Add(i);
             }
 
-            // TODO: Complete member initialization
-            this.shiftRegister = shiftRegister;
-
-            mfPin1ComboBox.SelectedValue = byte.Parse(shiftRegister.LatchPin);
-            mfPin2ComboBox.SelectedValue = byte.Parse(shiftRegister.ClockPin);
-            mfPin3ComboBox.SelectedValue = byte.Parse(shiftRegister.DataPin);
+            ////mfPin1ComboBox.SelectedValue = byte.Parse(shiftRegister.LatchPin);
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFShiftRegisterPanel.cs:47]: simply remove.

File: UI/Panels/Device/MFShiftRegisterPanel.cs [hunk 6]
```diff
@@ -29,52 +25,75 @@ public MFShiftRegisterPanel()
...
-                mfPin2ComboBox.SelectedIndex = 1;
-                mfPin3ComboBox.SelectedIndex = 2;                
-            }
+            this.shiftRegister = shiftRegister;
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin1ComboBox, Pins, shiftRegister.LatchPin);
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin2ComboBox, Pins, shiftRegister.ClockPin);
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin3ComboBox, Pins, shiftRegister.DataPin);
+            update_lists();
+
+            ////if (mfPin1ComboBox.Items.Count > 2)
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFShiftRegisterPanel.cs:36]: simply remove.

File: UI/Panels/Device/MFShiftRegisterPanel.cs [hunk 7]
```diff
@@ -29,52 +25,75 @@ public MFShiftRegisterPanel()
...
+            pinList = Pins; // Keep pin list stored
 
-            if (mfPin1ComboBox.Items.Count > 2)
-            {
-                mfPin1ComboBox.SelectedIndex = 0;
-                mfPin2ComboBox.SelectedIndex = 1;
-                mfPin3ComboBox.SelectedIndex = 2;                
-            }
+            this.shiftRegister = shiftRegister;
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin1ComboBox, Pins, shiftRegister.LatchPin);
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFShiftRegisterPanel.cs:31]: simply remove.

File: UI/Panels/Device/MFServoPanel.cs [hunk 8]
```diff
@@ -1,54 +1,42 @@
...
             ComboBoxHelper.BindMobiFlightFreePins(mfPinComboBox, Pins, servo.DataPin);
 
-            if (mfPinComboBox.Items.Count > 0)
-            {
-                mfPinComboBox.SelectedIndex = 0;
-            }
-            // TODO: Complete member initialization
-            this.servo = servo;
-            mfPinComboBox.SelectedValue = byte.Parse(servo.DataPin);
+            ////if (mfPinComboBox.Items.Count > 0) {
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFServoPanel.cs:24]: simply remove.

File: UI/Panels/Device/MFOutputPanel.cs [hunk 9]
```diff
@@ -11,47 +11,41 @@ namespace MobiFlight.UI.Panels.Settings.Device
...
-            if (mfPinComboBox.Items.Count > 0)
-            {
-                mfPinComboBox.SelectedIndex = 0;
-            }
-
-            // TODO: Complete member initialization
-            this.output = output;
-            mfPinComboBox.SelectedValue = byte.Parse(output.Pin);
+            ComboBoxHelper.BindMobiFlightFreePins(mfPinComboBox, Pins, output.Pin);
+            ////if (mfPinComboBox.Items.Count > 0) {
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFOutputPanel.cs:33]: simply remove.

File: UI/Panels/Device/MFLedSegmentPanel.cs [hunk 10]
```diff
@@ -27,53 +26,71 @@ public MFLedSegmentPanel()
...
+            ledModule.Name = textBox1.Text;
+            ledModule.Brightness = (byte)(mfIntensityTrackBar.Value);
+            ledModule.NumModules = mfNumModulesComboBox.Text;
+        }
+        private void update_lists()
+        {
+            bool ex_initialized = initialized;
+            initialized = false;    // inhibit value_Changed events
+            ComboBoxHelper.BindMobiFlightFreePins(mfPin1ComboBox, pinList, ledModule.DinPin);
+            //mfPin1ComboBox.SelectedValue = byte.Parse(ledModule.DinPin);  // moved inside BindMobiFlightFreePins()
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFLedSegmentPanel.cs:63]: simply remove.

File: UI/Panels/Device/MFLedSegmentPanel.cs [hunk 11]
```diff
@@ -27,53 +26,71 @@ public MFLedSegmentPanel()
...
+            ComboBoxHelper.BindMobiFlightFreePins(mfPin2ComboBox, pinList, ledModule.ClsPin);
+            //mfPin2ComboBox.SelectedValue = byte.Parse(ledModule.ClsPin);  // moved inside BindMobiFlightFreePins()
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFLedSegmentPanel.cs:65]: simply remove.

File: UI/Panels/Device/MFLedSegmentPanel.cs [hunk 12]
```diff
@@ -27,53 +26,71 @@ public MFLedSegmentPanel()
...
+            ComboBoxHelper.BindMobiFlightFreePins(mfPin3ComboBox, pinList, ledModule.ClkPin);
+            //mfPin3ComboBox.SelectedValue = byte.Parse(ledModule.ClkPin);  // moved inside BindMobiFlightFreePins()
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFLedSegmentPanel.cs:67]: simply remove.

File: UI/Panels/Device/MFLedSegmentPanel.cs [hunk 13]
```diff
@@ -27,53 +26,71 @@ public MFLedSegmentPanel()
...
-            // TODO: Complete member initialization
             this.ledModule = ledModule;
 
-            mfPin1ComboBox.SelectedValue = byte.Parse(ledModule.DinPin);
-            mfPin2ComboBox.SelectedValue = byte.Parse(ledModule.ClsPin);
-            mfPin3ComboBox.SelectedValue = byte.Parse(ledModule.ClkPin);
+            update_lists();
 
-            ComboBoxHelper.SetSelectedItem(mfNumModulesComboBox, ledModule.NumModules);
+            //if (mfPin1ComboBox.Items.Count > 2) {
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFLedSegmentPanel.cs:40]: simply remove.

File: UI/Panels/Device/MFLedSegmentPanel.cs [hunk 14]
```diff
@@ -27,53 +26,71 @@ public MFLedSegmentPanel()
             if (Parent != null) mfIntensityTrackBar.BackColor = Parent.BackColor;
         }
 
-        public MFLedSegmentPanel(MobiFlight.Config.LedModule ledModule, List<MobiFlightPin> Pins):this()
+        public MFLedSegmentPanel(MobiFlight.Config.LedModule ledModule, List<MobiFlightPin> Pins) : this()
         {
-            ComboBoxHelper.BindMobiFlightFreePins(mfPin1ComboBox, Pins, ledModule.DinPin);
-            ComboBoxHelper.BindMobiFlightFreePins(mfPin2ComboBox, Pins, ledModule.ClsPin);
-            ComboBoxHelper.BindMobiFlightFreePins(mfPin3ComboBox, Pins, ledModule.ClkPin);
+            pinList = Pins; // Keep pin list stored
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFLedSegmentPanel.cs:31]: put the comment on top/before

File: UI/Panels/Device/MFInputShiftRegisterPanel.cs [hunk 15]
```diff
@@ -29,51 +30,76 @@ public MFInputShiftRegisterPanel()
...
-                mfPin1ComboBox.SelectedIndex = 0;
-                mfPin2ComboBox.SelectedIndex = 1;
-                mfPin3ComboBox.SelectedIndex = 2;
-            }
+            pinList = Pins; // Keep pin list stored
 
-            for (int i = 1; i <= MAX_MODULES; i++)
-            {
+            this.inputShiftRegister = inputShiftRegister;
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin1ComboBox, Pins, inputShiftRegister.LatchPin);
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFInputShiftRegisterPanel.cs:36]: simply remove.

File: UI/Panels/Device/MFInputShiftRegisterPanel.cs [hunk 16]
```diff
@@ -29,51 +30,76 @@ public MFInputShiftRegisterPanel()
...
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin2ComboBox, Pins, inputShiftRegister.ClockPin);
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfPin3ComboBox, Pins, inputShiftRegister.DataPin);
+            update_lists();
+
+            ////if (mfPin1ComboBox.Items.Count > 2) {
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFInputShiftRegisterPanel.cs:41]: simply remove.

File: UI/Panels/Device/MFInputShiftRegisterPanel.cs [hunk 17]
```diff
@@ -29,51 +30,76 @@ public MFInputShiftRegisterPanel()
...
+            for (int i = 1; i <= MAX_MODULES; i++) {
                 mfNumModulesComboBox.Items.Add(i);
             }
 
-            this.inputShiftRegister = inputShiftRegister;
-
-            ComboBoxHelper.SetSelectedItem(mfPin1ComboBox, inputShiftRegister.LatchPin);
-            ComboBoxHelper.SetSelectedItem(mfPin2ComboBox, inputShiftRegister.ClockPin);
-            ComboBoxHelper.SetSelectedItem(mfPin3ComboBox, inputShiftRegister.DataPin);
+            ////ComboBoxHelper.SetSelectedItem(mfPin1ComboBox, inputShiftRegister.LatchPin);
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFInputShiftRegisterPanel.cs:51]: simply remove.

File: UI/Panels/Device/MFEncoderPanel.cs [hunk 18]
```diff
@@ -26,47 +25,70 @@ public MFEncoderPanel()
...
+            //setNonPinValues();
 
             initialized = true;
         }
 
-        private void value_Changed(object sender, EventArgs e)
+        private void setNonPinValues()
         {
-            if (!initialized) return;
+            ////encoder.PinLeft = mfLeftPinComboBox.SelectedItem.ToString();
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFEncoderPanel.cs:54]: simply remove.

File: UI/Panels/Device/MFEncoderPanel.cs [hunk 19]
```diff
@@ -26,47 +25,70 @@ public MFEncoderPanel()
...
+            ////encoder.PinRight = mfRightPinComboBox.SelectedItem.ToString();
+            encoder.EncoderType = mfEncoderTypeComboBox.SelectedIndex.ToString();
+            encoder.Name = textBox1.Text;
+        }
+        private void update_lists()
+        {
+            bool ex_initialized = initialized;
+            initialized = false;    // inhibit value_Changed events
+            ComboBoxHelper.BindMobiFlightFreePins(mfLeftPinComboBox, pinList, encoder.PinLeft);
+            //mfLeftPinComboBox.SelectedValue = byte.Parse(encoder.PinLeft);  // moved inside BindMobiFlightFreePins()
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFEncoderPanel.cs:64]: simply remove.

File: UI/Panels/Device/MFEncoderPanel.cs [hunk 20]
```diff
@@ -26,47 +25,70 @@ public MFEncoderPanel()
...
+            ComboBoxHelper.BindMobiFlightFreePins(mfRightPinComboBox, pinList, encoder.PinRight);
+            //mfRightPinComboBox.SelectedValue = byte.Parse(encoder.PinRight);  // moved inside BindMobiFlightFreePins()
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFEncoderPanel.cs:66]: simply remove.

File: UI/Panels/Device/MFEncoderPanel.cs [hunk 21]
```diff
@@ -26,47 +25,70 @@ public MFEncoderPanel()
...
-            : this()
+        public MFEncoderPanel(MobiFlight.Config.Encoder encoder, List<MobiFlightPin> Pins) : this()
         {
-            ComboBoxHelper.BindMobiFlightFreePins(mfLeftPinComboBox, Pins, encoder.PinLeft);
-            ComboBoxHelper.BindMobiFlightFreePins(mfRightPinComboBox, Pins, encoder.PinRight);
-            
+            pinList = Pins;
+            this.encoder = encoder;
+            update_lists();
+            ////ComboBoxHelper.BindMobiFlightFreePins(mfLeftPinComboBox, Pins, encoder.PinLeft);
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFEncoderPanel.cs:33]: simply remove.

File: UI/Panels/Device/MFAnalogPanel.cs [hunk 22]
```diff
@@ -40,33 +36,30 @@ private void SensitivityTrackBar_ValueChanged(object sender, EventArgs e)
...
-                mfPinComboBox.SelectedIndex = 0;
-            }
-
             this.analog = analogDevice;
-            mfPinComboBox.SelectedValue = byte.Parse(analog.Pin);
-            SensitivityTrackBar.Value = byte.Parse(analog.Sensitivity);
 
-            textBox1.Text = analog.Name;
-            setValues();
+            ////if (mfPinComboBox.Items.Count > 0)
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFAnalogPanel.cs:46]: simply remove.

File: UI/Panels/Device/MFAnalogPanel.cs [hunk 23]
```diff
@@ -40,33 +36,30 @@ private void SensitivityTrackBar_ValueChanged(object sender, EventArgs e)
...
+            ////{
+            ////    mfPinComboBox.SelectedIndex = 0;
+            ////}
 
+            mfPinComboBox.SelectedValue = byte.Parse(analog.Pin);
+            textBox1.Text = analog.Name;
+            SensitivityTrackBar.Value = byte.Parse(analog.Sensitivity);
+            
+            //setValues();
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFAnalogPanel.cs:55]: simply remove.

File: UI/Panels/Device/MFLedSegmentPanel.cs [hunk 24]
```diff
@@ -27,53 +26,71 @@ public MFLedSegmentPanel()
...
         }
+        private void setNonPinValues()
+        {
+            ledModule.Name = textBox1.Text;
+            ledModule.Brightness = (byte)(mfIntensityTrackBar.Value);
+            ledModule.NumModules = mfNumModulesComboBox.Text;
+        }
+        private void update_lists()
+        {
+            bool ex_initialized = initialized;
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFLedSegmentPanel.cs:60]: no underscrore, camelCase is preferred

File: UI/Panels/Settings/MobiFlightPanel.cs [hunk 25]
```diff
@@ -150,27 +150,17 @@ public void SaveSettings()
...
-            TreeNode parentNode = this.mfModulesTreeView.SelectedNode;
-            if (parentNode == null) return;
-
-            if (this.mfModulesTreeView.SelectedNode == null) return;
-
-            while (parentNode.Level > 0) parentNode = parentNode.Parent;
-
-            MobiFlightModule module = parentNode.Tag as MobiFlightModule;
-
+            TreeNode moduleNode = getModuleNode();
```
Comment (DocMoebiuz) [inline UI/Panels/Settings/MobiFlightPanel.cs:153]: why are all these changes part of this PR?

Comment (DocMoebiuz) [reply UI/Panels/Settings/MobiFlightPanel.cs:153]: it looks like because of the renaming.

**Review by DocMoebiuz — CHANGES_REQUESTED** [review]
Thanks for all the effort. It seems to be working.

Improvements:
* It is really hard to review the changes because sometimes, some lines were only moved around, and indenting got changed - (e.g. MFAnalogPanel.cs). 
* There are also many commented lines of code. Please remove those since we have git for history. 
* Please don't put comments at the end of the line. Always better on top, before the code.
* Also for some reason it seems that the git history at least on my local main branch is not linear anymore. When you merge the main branch, do you merge rebase  or simply merge?
* Method normally are CamelCase.
* EventHandlers in C# have an "_eventname" when the Designer adds them, so that's why some methods contain an "_"

**Comment by GioCC** [comment]
Thanks for the indications; most of the requested changes were awaiting cleanup pending your review feedback (that's why the PR was in draft state). I will proceed to correct as per request.

**Comment by DocMoebiuz** [comment]
@GioCC - cool, thanks!!!

**Comment by GioCC** [comment]
> * It not easy to review the changes because sometimes, some lines were only moved around, and indenting got changed - (e.g. MFAnalogPanel.cs).

I think this is sometimes caused by different editor settings (eg tab lengths and the like); lines moving was probably in the attempt to make identical code more uniform in different places (or the flow more meaningful), although obviously I'd need to see where it occurred.

> * Please don't put comments at the end of the line. Always better on top, before the code.

I would normally disagree :) but of course guidelines are guidelines. Duly noted. 

> * Also for some reason it seems that the git history at least on my local main branch is not linear anymore. When you merge the main branch, do you merge rebase  or simply merge?

No, I usually simply merge by habit; I will definitely try and be more careful in the future.

> * Method normally are CamelCase.

Can't find anymore where I did differently, can you still recall some pointers so I can correct?

> * EventHandlers in C# have an "_eventname" when the Designer adds them, so that's why some methods contain an "_"

Ooooohhh... ok, I was scratching my head to try and remember what this was about, until I copied the text in the editor to reply... the underscore appeared, and suddenly everything made sense :D

**Comment by GioCC** [comment]
Is there any other change required?

**Comment by GioCC** [comment]
I reviewed the changes after your prompt, I believe almost everything is now in place, except possibly:
- there are still some (very short) comments on the end of the line; it did not look like a very big issue for this PR, so I left them, taking the notice for future modifications. If you think that's critical, I can go ahead and correct them.
- I did not try and correct spacing differences; that would probably require retracing the whole changes and reapply them one by one on the starting code base. Same considerations as above.
That's why I requested your review again, to confirm whether the points above can be accepted.

File: UI/Panels/Device/MFShiftRegisterPanel.cs [hunk 26]
```diff
@@ -29,52 +25,59 @@ public MFShiftRegisterPanel()
...
+            pinList = Pins; // Keep pin list stored
 
-            if (mfPin1ComboBox.Items.Count > 2)
-            {
-                mfPin1ComboBox.SelectedIndex = 0;
-                mfPin2ComboBox.SelectedIndex = 1;
-                mfPin3ComboBox.SelectedIndex = 2;                
-            }
+            this.shiftRegister = shiftRegister;
+            update_lists();
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFShiftRegisterPanel.cs:31]: Let's change this to a more speaking name and use camel case:
 UpdateFreePinsInDropDowns()

File: UI/Panels/Device/MFShiftRegisterPanel.cs [hunk 27]
```diff
@@ -29,52 +25,59 @@ public MFShiftRegisterPanel()
...
         }
 
-        private void value_Changed(object sender, EventArgs e)
+        private void setNonPinValues()
         {
-            if (!initialized) return;
+            shiftRegister.Name = textBox1.Text;
+            shiftRegister.NumModules = string.IsNullOrEmpty(mfNumModulesComboBox.Text) ? "1" : mfNumModulesComboBox.Text;
+        }
+        private void update_lists()
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFShiftRegisterPanel.cs:48]: Let's change this to a more speaking name and use camel case:
UpdateFreePinsInDropDowns()

File: UI/Panels/Device/MFShiftRegisterPanel.cs [hunk 28]
```diff
@@ -29,52 +25,59 @@ public MFShiftRegisterPanel()
...
+            bool ex_initialized = initialized;
+            initialized = false;    // inhibit value_Changed events
+            ComboBoxHelper.BindMobiFlightFreePins(mfPin1ComboBox, pinList, shiftRegister.LatchPin);
+            ComboBoxHelper.BindMobiFlightFreePins(mfPin2ComboBox, pinList, shiftRegister.ClockPin);
+            ComboBoxHelper.BindMobiFlightFreePins(mfPin3ComboBox, pinList, shiftRegister.DataPin);
+            initialized = ex_initialized;
+        }
 
-            setValues();
+        private void update_all(ComboBox comboBox)
```
Comment (DocMoebiuz) [inline UI/Panels/Device/MFShiftRegisterPanel.cs:58]: Let's change this to a more speaking name and use camel case:
ReassignFreePinsInDropDowns()

**Review by DocMoebiuz — CHANGES_REQUESTED** [review]
Ok, thanks so much for all your efforts.
Last tiny adjustment would be the naming for the 

update_lists() > UpdateFreePinsInDropDowns()
update_all() > RefreshFreePinsInDropDowns()

**Pull request merged**